types/index.d.ts

```diff
@@ -0,0 +1,496 @@
+    ? Models[K] extends ModelDefinition<infer Data>
+      ? FlattenRelationships<Data>
+      : {}
+    : {};
+
+  // Extracts factory definition info for the given key, if a corresponding factory is defined
+  type ExtractFactoryData<Factories, K> = K extends keyof Factories
+    ? Factories[K] extends FactoryDefinition<infer Data>
+      ? FlattenFactoryMethods<Data>
+      : {}
+    : {};
+
+  /**
+   * Models all available information about a given set of model and
+   * factory definitions, determining the behavior of ORM methods on
+   * a `Server` and its corresponding `Schema` instance.
+   */
+  export type Registry<
+    Models extends Record<string, ModelDefinition>,
+    Factories extends Record<string, FactoryDefinition>
+  > = {
+    [K in keyof Models | keyof Factories]: ModelInstance<
+      ExtractModelData<Models, K> & ExtractFactoryData<Factories, K>
+    >;
+  };
+
+  /** Represents the type of an instantiated Mirage model.  */
+  export type ModelInstance<Data = {}> = Data & {
+    id?: string;
+    attrs: Record<string, unknown>;
+    modelName: string;
+
+    /** Persists any updates on this model back to the Mirage database. */
+    save(): void;
+
+    /** Updates and immediately persists a single or multiple attr(s) on this model. */
+    update<K extends keyof Data>(key: K, value: Data[K]): void;
+    update(changes: Partial<Data>): void;
+
+    /** Removes this model from the Mirage database. */
+    destroy(): void;
+
+    /** Reloads this model's data from the Mirage database. */
+    reload(): void;
+  };
+}
+
+declare module "miragejs/server" {
+  import { Request, Response } from "miragejs";
+  import { ModelInstance } from "miragejs/-types";
+  import Db from "miragejs/db";
+  import IdentityManager from "miragejs/identity-manager";
+  import Schema from "miragejs/orm/schema";
+  import PretenderServer from "pretender";
+
+  type MaybePromise<T> = T | PromiseLike<T>;
+
+  /** A callback that will be invoked when a given Mirage route is hit. */
+  export type RouteHandler<Registry> = (
+    schema: Schema<Registry>,
+    request: Request
+  ) => MaybePromise<ModelInstance | Response | object>;
+
+  export interface HandlerOptions {
+    /** A number of ms to artificially delay responses to this route. */
+    timing?: number;
+  }
+
+  export interface ServerConfig {
+    urlPrefix?: string;
+    namespace?: string;
+    timing?: number;
+    environment?: string;
+    trackRequests?: boolean;
+    useDefaultPassthroughs?: boolean;
+    logging?: boolean;
+
+    seeds?: (server: Server) => void;
+    scenarios?: (server: Server) => void;
+
+    routes?: (this: Server) => void;
+    baseConfig?: (this: Server) => void;
+    testConfig?: (this: Server) => void;
+
+    inflector?: object;
+    identityManagers?: IdentityManager;
+    models?: any;
+    serializers?: any;
+    factories?: any;
+
+    pretender?: PretenderServer;
+  }
+
+  export class Server<Registry = Record<string, ModelInstance>>
```

Comment:
Should this be constrained as to what it extends, or no? If so:

```suggestion
  export class Server<Registry extends Record<string, ModelInstance> = Record<string, ModelInstance>>
```

Reply:
You right, it is better if there is a type constrain. Thanks.